rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+type storage interface {
+	// ReadLastProcessedEvent reads the last processed event data
+	ReadLastProcessedEvent(string) (string, bool)
+
+	// WriteLastProcessedEvent writes the last processed event data
+	WriteLastProcessedEvent(data string, contractAddr string) error
+}
+
+// eventData holds information on event data mapping
+type eventData struct {
+	payloadType  rootchain.PayloadType
+	eventABI     *abi.Event
+	methodABI    *abi.Method
+	localAddress types.Address
+}
+
+// SAMUEL is the module that coordinates activities with the SAMP and Event Tracker
+type SAMUEL struct {
+	eventData eventData
+	logger    hclog.Logger
+
+	eventTracker eventTracker
+	samp         samp
+	storage      storage
+	signer       signer
+	transport    transport
+}
+
+// NewSamuel creates a new SAMUEL instance
+func NewSamuel(
+	configEvent *rootchain.ConfigEvent,
+	logger hclog.Logger,
+	eventTracker eventTracker,
+	samp samp,
+	signer signer,
+	storage storage,
+	transport transport,
+) *SAMUEL {
+	return &SAMUEL{
+		logger:       logger.Named("SAMUEL"),
+		eventData:    initEventData(configEvent),
+		eventTracker: eventTracker,
+		samp:         samp,
+		signer:       signer,
+		storage:      storage,
+		transport:    transport,
+	}
+}
+
+// initEventLookupMap generates the SAMUEL event data lookup map from the
+// passed in rootchain configuration
+func initEventData(
+	configEvent *rootchain.ConfigEvent,
+) eventData {
+	return eventData{
+		payloadType:  configEvent.PayloadType,
+		eventABI:     abi.MustNewEvent(configEvent.EventABI),
+		methodABI:    abi.MustNewABI(configEvent.MethodABI).GetMethod(configEvent.MethodName),
+		localAddress: types.StringToAddress(configEvent.LocalAddress),
+	}
+}
+
+// Start starts the SAMUEL module
+func (s *SAMUEL) Start() error {
+	// Start the event loop for the tracker
+	s.startEventLoop()
+
+	// Register the gossip message handler
+	if err := s.registerGossipHandler(); err != nil {
+		return fmt.Errorf("unable to register gossip handler, %w", err)
+	}
+
+	// Fetch the latest event data
+	startBlock, err := s.getStartBlockNumber()
+	if err != nil {
+		return fmt.Errorf("unable to get start block number, %w", err)
+	}
+
+	// Start the Event Tracker
+	if err := s.eventTracker.Start(startBlock); err != nil {
+		return fmt.Errorf("unable to start event tracker, %w", err)
+	}
+
+	return nil
+}
+
+// getStartBlockNumber determines the starting block for the Event Tracker
+func (s *SAMUEL) getStartBlockNumber() (uint64, error) {
+	startBlock := rootchain.LatestRootchainBlockNumber
+
+	data, exists := s.storage.ReadLastProcessedEvent(s.eventData.localAddress.String())
+	if exists && data != "" {
+		// index:blockNumber
+		values := strings.Split(data, ":")
+		if len(values) < 2 {
+			return 0, fmt.Errorf("invalid last processed event in DB: %v", values)
+		}
+
+		blockNumber, err := strconv.ParseUint(values[1], 10, 64)
+		if err != nil {
+			return 0, fmt.Errorf("unable to parse last processed block number in DB: %w", err)
+		}
```

Comment:
How about defining functions to parse and generate the date to be stored in storage?

Reply:
The snippet you linked is already part of the helper method.

Saving to the DB is a one-liner